File: 450-Delete_Node_in_a_BST/main.go
package main

type TreeNode struct {
	Val int
	Left *TreeNode
	Right *TreeNode
}

func deleteNode(root *TreeNode, key int) *TreeNode {
	if root == nil {
		return nil
	}

	if root.Val > key {
		root.Left = deleteNode(root.Left, key)
		return root
	} else if root.Val < key {
		root.Right = deleteNode(root.Right, key)
		return root
	} else {
	if root.Right == nil {
		left := root.Left
		root.Left = nil
		return left
	}

	if root.Left == nil {
		right := root.Right
		root.Right = nil
		return right
	}

	successor := getMinNode(root.Right)
	successor.Left = root.Left
	successor.Right = removeMinNode(root.Right)
	root.Left = nil
	root.Right = nil
	return successor
	}
}

func getMinNode(node *TreeNode) *TreeNode {
	if node.Left == nil {
		return node
	}

	return getMinNode(node.Left)
}

func removeMinNode(node *TreeNode) *TreeNode {
	if node.Left == nil {
		right := node.Right
		node.Right = nil
		return right
	}

	node.Left = removeMinNode(node.Left)
	return node
}

func main() {

}

File: 450-Delete_Node_in_a_BST/main.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} key
 * @return {TreeNode}
 */

function TreeNode(val) {
  this.val = val;
  this.left = this.right = null;
};

var deleteNode = function (root, key) {
  if (root == null) {
    return null;
  }

  if (root.val > key) {
    root.left = deleteNode(root.left, key);
    return root;
  } else if (root.val < key) {
    root.right = deleteNode(root.right, key);
    return root;
  } else {

    // root.key === key, root 即被删除的那个节点
    // 没有右子树
    if (root.right === null) {
      let left = root.left;
      root.left = null;
      return left;
    }

    // 没有左子树
    if (root.left === null) {
      let right = root.right;
      root.right = null;
      return right;
    }

    // 两颗子树都在
    let successor = getMinNode(root.right);
    let left = root.left;
    let right = removeMinNode(root.right);
    successor.left = left;
    successor.right = right;
    root.left = root.right = null;
    return successor;
  }
};

function getMinNode(node) {
  if (node.left === null) {
    return node;
  }
  return getMinNode(node.left);
}

function removeMinNode(node) {
  if (node.left === null) {
    let right = node.right;
    node.right = null;
    return right;
  }

  node.left = removeMinNode(node.left);
  return node;
}

// let two = new TreeNode(2);
// let four = new TreeNode(4);
// let three = new TreeNode(3);
// three.left = two;
// three.right = four;
// let seven = new TreeNode(7);
// let six = new TreeNode(6);
// six.right = seven;
// let root = new TreeNode(5);
// root.left = three;
// root.right = six;

// // console.log(root);
// deleteNode(root, 3);
// console.log(root);